lib/bundle/defaults.js

```diff
@@ -19,75 +19,163 @@ function getGenericDefaults (generator) {
         let existingGeneratedKey = generator.getExistingGeneratedKey(schema, file);
 
         if (existingGeneratedKey === undefined) {
-          existingGeneratedKey = generator.generateKeyForFilepath(schema, file);
+          existingGeneratedKey = generator.generateKeyForFilepath(schema, file, pathFromRoot);
         }
 
         if (existingGeneratedKey === null) {
           return null;
         }
 
-        if (!generator.isInRoot(hash)) {
+        if (!generator.isInRoot(hash, pathFromRoot)) {
           return null;
         }
 
-        return generator.generateKeyForPointer(schema, existingGeneratedKey + hash.slice(1));
+        return generator.generateKeyForPointer(schema, existingGeneratedKey + hash.slice(1), pathFromRoot);
       }
 
       if (url.isHttp(file)) {
-        return generator.generateKeyForUrl(schema, file);
+        return generator.generateKeyForUrl(schema, file, pathFromRoot);
       }
 
-      return generator.generateKeyForFilepath(schema, file);
+      return generator.generateKeyForFilepath(schema, file, pathFromRoot);
     },
   };
 }
 
 module.exports.getGenericDefaults = getGenericDefaults;
 
-module.exports.getDefaultsForOldJsonSchema = function (defaults = getGenericDefaults(new KeyGenerator("#/definitions"))) {
+module.exports.getDefaultsForOldJsonSchema = function (defaults = getGenericDefaults(new KeyGenerator(() => "#/definitions"))) {
   return defaults;
 };
 
-module.exports.getDefaultsForNewJsonSchema = function (defaults = getGenericDefaults(new KeyGenerator("#/$defs"))) {
+module.exports.getDefaultsForNewJsonSchema = function (defaults = getGenericDefaults(new KeyGenerator(() => "#/$defs"))) {
   return defaults;
 };
 
-module.exports.getDefaultsForOAS2 = function (defaults = getGenericDefaults(new KeyGenerator("#/definitions"))) {
+const defaultOas2RootResolver = (pathFromRoot) => {
+  const pathParsed = pathFromRoot.split("/");
+
+  if (canReferenceParameter(pathParsed)) {
+    return "#/parameters";
+  }
+  if (canReferenceResponse(pathParsed)) {
+    return "#/responses";
+  }
+
+  return "#/definitions";
```

Comment:
hmm, perhaps let's return null (where null = no custom root, cannot detect).
Obviously will need to add another condition for a schema.